mesop/labs/chat.py

```diff
@@ -0,0 +1,228 @@
+_STYLE_CHAT_INPUT_BOX = me.Style(padding=me.Padding(top="30px"))
+_STYLE_CHAT_BUBBLE_NAME = me.Style(
+  font_weight="bold",
+  font_size="12px",
+  padding=me.Padding(left="15px", right="15px", bottom="5px"),
+)
+_STYLE_CHAT_BUBBLE_PLAINTEXT = me.Style(
+  margin=me.Margin(top="15px", bottom="15px")
+)
+
+
+def _make_style_chat_ui_container(has_title: bool) -> me.Style:
+  """Generates styles for chat UI container depending on if there is a title or not.
+
+  Args:
+    has_title: Whether the Chat UI is display a title or not.
+  """
+  return me.Style(
+    display="grid",
+    grid_template_columns="repeat(1, 1fr)",
+    grid_template_rows="1fr 14fr 1fr" if has_title else "5fr 1fr",
+    margin=me.Margin(top=0, bottom=0, left="auto", right="auto"),
+    width="min(1024px, 100%)",
+    height="100vh",
+    background="#fff",
+    box_shadow=(
+      "0 3px 1px -2px #0003, 0 2px 2px #00000024, 0 1px 5px #0000001f"
+    ),
+    padding=_DEFAULT_PADDING,
+  )
+
+
+def _make_style_chat_bubble_wrapper(
+  role: Literal["user", "assistant"]
+) -> me.Style:
+  """Generates styles for chat bubble position.
+
+  Args:
+    role: Chat bubble alignment depends on the role
+  """
+  align_items = "end" if role == _ROLE_USER else "start"
+  return me.Style(
+    display="flex",
+    flex_direction="column",
+    align_items=align_items,
+  )
+
+
+def _make_chat_bubble_style(role: Literal["user", "assistant"]) -> me.Style:
+  """Generates styles for chat bubble.
+
+  Args:
+    role: Chat bubble background color depends on the role
+  """
+  background = (
+    _COLOR_CHAT_BUBBLE_YOU if role == _ROLE_USER else _COLOR_CHAT_BUBBLE_BOT
+  )
+  return me.Style(
+    width="80%",
+    font_size="13px",
+    background=background,
+    border_radius="15px",
+    padding=me.Padding(
+      top="0px",
+      right="15px",
+      left="15px",
+      bottom="3px",
+    ),
+    margin=me.Margin(bottom="10px"),
+    border=me.Border(
+      left=_DEFAULT_BORDER_SIDE,
+      right=_DEFAULT_BORDER_SIDE,
+      top=_DEFAULT_BORDER_SIDE,
+      bottom=_DEFAULT_BORDER_SIDE,
+    ),
+  )
+
+
+@dataclass
+class Message:
+  """Chat message metadata."""
+
+  role: Literal["assistant", "user"] = "user"
+  username: str = ""
+  content: str = ""
+
+
+@me.stateclass
+class State:
+  input: str
+  output: list[Message]
+  in_progress: bool = False
+
+
+def on_input_update(State):
+  """Generic on text input handler that saves input to State using the given key.
+
+  This helper only works if you have one state instance. If use multiple state classes
+  with this helper, then only the last event handler will be stored. For more info, see
+  https://google.github.io/mesop/guides/troubleshooting/#avoid-using-closure-variables-in-event-handler.
+  """
+
+  def _on_update(e: me.InputEvent):
+    state = me.state(State)
+    setattr(state, e.key, e.value)
+
+  return _on_update
+
+
+def chat(
+  transform: Callable[[str], Generator[str, None, None] | str],
```

Comment:
I think transform should have two parameters:
1. prompt/input: str
2. history: list[ChatMessage] - it could potentially be a list of tuples like how [Gradio](https://www.gradio.app/guides/creating-a-chatbot-fast#defining-a-chat-function) does it, but I think defining a dataclass is more robust (e.g. if we need a new field , it won't break existing apps).

Reply:
Hmm. I agree that providing a history is a good idea. One downside of using a dataclass is that it forces the implementor of the transform function to account for Mesop specific behavior. So they'd need to add some light translation of the history to work with their LLM API whether formatting history into the prompt is necessary. But Gradio also has the same problem using primitives.

So I agree that a dataclass seems better.

Updated.

Reply:
I think Mesop developers doing some translation is a bit inevitable because I've seen a wide range of how history is formatted across internal and external APIs.